### ISSUE TITLE
initrd secrets should be per-generation

### ISSUE THREAD

**Issue by emilazy** [NixOS/nixpkgs#85563]
Currently, secrets are copied into initrds for every generation on every rebuild. This means that if you rotate initrd SSH host keys or change the layout of your initrd secrets directory, rollbacks are effectively broken.

I think that secrets should be copied into `/var/lib/initrd-secrets/<generation-id>` (where `<generation-id>` is some stable identifier for the system; generation number might be good enough here) on the `nixos-rebuild` of that generation, and the bootloader modules would copy from the corresponding secrets generation. This would also let us handle remapping the paths to the initrd form, adjusting permissions, etc. in one place, reducing code duplication and making it easier to fix #85000.

cc @lopsided98; what do you think of this design? If people agree that this is a good direction to take then I might look into drafting an implementation.

**Comment by lopsided98** [comment]
I like this idea and your proposed implementation.

**Comment by emilazy** [comment]
still relevant

**Comment by andrew-d** [comment]
still relevant

**Comment by pwaller** [comment]
I may have addressed this here, please take a look: https://github.com/NixOS/nixpkgs/pull/209156

**Comment by emilazy** [comment]
That PR seems GRUB-specific, unless I'm missing something; you can encounter this issue using other initrd-secrets-supporting bootloaders like systemd-boot, so unfortunately I don't think it'd solve this issue entirely.

**Comment by pwaller** [comment]
Right. If someone wanted to fix systemd too it looks like an analogous fix could be had fairly easily by modifying this in a similar way:

https://github.com/NixOS/nixpkgs/blob/af0b453724a59bc8fa90fa1dfad30896cc3df529/nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py#L109-L116

**Comment by emilazy** [comment]
Right; I was hoping the logic could be factored out to be independent of bootloaders to reduce duplication like this and help fix #85000, but fixing it in the individual bootloaders is still better than the status quo. Thanks for looking into this for GRUB.
